File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) 
    {
        int max_len = 0;
        int start = 0;
        map <char, int> mp;
        for(int end = 0; end < s.size(); end++)
        {
            if(mp.find(s[end])!=mp.end() && mp[s[end]] >= start)
            {
                start = mp[s[end]] + 1;
            }
            mp[s[end]] = end;
            max_len = max(max_len, end - start + 1);
        }
        
        return max_len;
        
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_len = 0 
        start, end = 0, 0
        dic = defaultdict(int)
        for end, c in enumerate(s):
            if c in dic:
                start = max(start, dic[c])
            max_len = max(max_len, end - start+1)
            dic[c] = end + 1
        
        return max_len
    
        """
        a : 0
        b : 1
        c : 2
        """